src/Workspaces/Core/Portable/Utilities/EditDistance.cs

```diff
@@ -186,24 +517,46 @@ public static bool IsCloseMatch(string originalText, string candidateText)
         /// </summary>
         public static bool IsCloseMatch(string originalText, string candidateText, out double matchCost)
         {
-            return IsCloseMatch(originalText, candidateText, GetCloseMatchThreshold(originalText), out matchCost);
+            using (var editDistance = new EditDistance(originalText))
+            {
+                return editDistance.IsCloseMatch(candidateText, out matchCost);
+            }
         }
 
-        public static int GetCloseMatchThreshold(string originalText)
+        public bool IsCloseMatch(string candidateText, out double matchCost)
         {
-            return Math.Min(4, originalText.Length / 2);
+            if (this.originalText.Length < 3)
+            {
+                // If we're comparing strings that are too short, we'll find 
+                // far too many spurious hits.  Don't even both in this case.
+                matchCost = double.MaxValue;
+                return false;
+            }
+
+            if (lastIsCloseMatchResult.Item1 == candidateText)
+            {
+                matchCost = lastIsCloseMatchResult.Item3;
+                return lastIsCloseMatchResult.Item2;
+            }
+
+            var result = IsCloseMatchWorker(candidateText, out matchCost);
+            lastIsCloseMatchResult = ValueTuple.Create(candidateText, result, matchCost);
+            return result;
         }
 
-        /// <summary>
-        /// Returns true if 'value1' and 'value2' are likely a misspelling of each other.
-        /// Returns false otherwise.  If it is a likely misspelling a matchCost is provided
-        /// to help rank the match.  Lower costs mean it was a better match.
-        /// </summary>
-        public static bool IsCloseMatch(string originalText, string candidateText, int costThreshold, out double matchCost)
+        private bool IsCloseMatchWorker(string candidateText, out double matchCost)
         {
-            matchCost = EditDistance.GetEditDistance(originalText, candidateText);
+            matchCost = double.MaxValue;
+
+            // If the two strings differ by more characters than the cost threshold, then there's 
+            // no point in even computing the edit distance as it would necessarily take at least
+            // that many additions/deletions.
+            if (Math.Abs(originalText.Length - candidateText.Length) <= threshold)
+            {
+                matchCost = GetEditDistance(candidateText);
```

Comment:
Which algorithms want the boolean and which want the actual distance? If enough just want the boolean, maybe we could inform `GetEditDistance` that we're looking for the boolean here and it could stop the search earlier.